tsne/bh_sne_src/tsne.cpp

```diff
@@ -54,10 +56,39 @@ using std::fprintf;
 using std::log;
 using std::move;
 using std::sqrt;
+using std::unique_ptr;
 using std::vector;
 
 namespace {
 
+struct TSNEState {
+ public:
+  bool step_by(int n);
+
+ private:
+  template <int NDIMS>
+  bool step_by_impl(int n);
+
+ public:
+  int N;
+  double* Y;
+  int no_dims;
+  double theta;
+  int max_iter;
+  int stop_lying_iter;
+  int mom_switch_iter;
+  int iter;
+  float total_time;
+  float residual_time;
+  std::vector<double> P;
+  std::vector<unsigned int> row_P;
+  std::vector<unsigned int> col_P;
+  std::vector<double> val_P;
+  std::vector<double> dY;
+  std::vector<double> uY;
+  std::vector<double> gains;
```

Comment:
Is all of this really part of the public API?  I'm a bit unconvinced.  Not that it matters really, since the C API is what gets exported and as far as it's concerned this is all an opaque blob.

Reply:
No, it's not public API. I can and should fix this.